### PR TITLE
Adding jackknife plus

### PR DESCRIPTION
**Pull request by paolo-vergo**
As requested I introduced Jackknife+ as integration to the original **conformal.pred.jack** function.

### CODE REVIEW

File: conformalInference/R/jack.R [hunk 1]
```diff
@@ -111,56 +135,141 @@ conformal.pred.jack = function(x, y, x0, train.fun, predict.fun, alpha=0.1,
...
+        mu=matrix(one.sapply(1:n, function(jj){
+          return(c(predict.fun(updated_models [[jj]],matrix(x0[ii,],nrow=1))))
+        }),nrow=n)  ## matrix of dim n x m
+        
+        
+        # Get quantiles
+        lo_vec<-one.sapply(1:m, function(l) mu[,l]-res[,l]*mad.x0.vec[[l]][,ii])
+        up_vec<-one.sapply(1:m, function(l) mu[,l]+res[,l]*mad.x0.vec[[l]][,ii])
+        lo.obs <- one.sapply(1:m, function(l) quantile(lo_vec[,l],alpha))
+        up.obs <- one.sapply(1:m, function(l) quantile(up_vec[,l],1-alpha))
```
Comment (ryantibs) [inline conformalInference/R/jack.R:232]: We should be using the proper corrections for finite-samples here as in the jackknife+ paper. For example `ceil((n+1)*(1-alpha))/n` for the upper quantile. And the lower one is similar, I think it is `floor((n+1)*alpha)/n`. But this is from memory: please double-check that paper to be sure.

Comment (ryantibs) [inline conformalInference/R/jack.R:232]: We should be using the proper corrections for finite-samples here as in the jackknife+ paper. For example `ceil((n+1)*(1-alpha))/n` for the upper quantile. And the lower one is similar, I think it is `floor((n+1)*alpha)/n`. But this is from memory: please double-check that paper to be sure.

**Review by ryantibs — COMMENTED** [review]
Thanks @paolo-vergo for implementing this and I'm really sorry for the long delay.

Looking good overall. I did leave a bunch of comments; some of them are minor, but some of them are not.

**Comment by ryantibs** [comment]
By the way, @paolo-vergo: please add yourself (and any others who contributed to this code) as authors in the DESCRIPTION file of this PR!

**Comment by paolo-vergo** [comment]
@ryantibs I have updated my branch, integrating all the corrections.

**Comment by ryantibs** [comment]
Thanks! Can you please respond to the comments raised above and resolve the conversations if you think they're resolved? That will help me get a better sense of where things stand.

Also, just to double check: are all the people you listed under authors contribute to the code itself? This is supposed to be a list of authors for the package development (not for the research ideas that support it).

File: conformalInference/R/common.R [hunk 2]
```diff
@@ -38,3 +38,23 @@ weighted.quantile = function(v, prob, w=NULL, sorted=FALSE) {
...
   else return(v[min(i)])
 }
+
+##Automatically switch between parallel or serial computation
+
+
+
+
+one.sapply= function(index, fun){
+  if("future.apply" %in% rownames(installed.packages()))
```
[the remark this answers is not in the document]
Comment (paolo-vergo) [reply conformalInference/R/common.R]: Done.

File: conformalInference/R/jack.R [hunk 3]
```diff
@@ -52,7 +52,10 @@
 #' @param alpha Miscoverage level for the prediction intervals, i.e., intervals
 #'   with coverage 1-alpha are formed. Default for alpha is 0.1.
 #' @param verbose Should intermediate progress be printed out? Default is FALSE.
-#'
+#' @param plus Should the Jaccknife+ version of the algorithm be used?
```
[the remark this answers is not in the document]
Comment (paolo-vergo) [reply conformalInference/R/jack.R]: Okay I will set it as default

File: conformalInference/R/jack.R [hunk 4]
```diff
@@ -111,56 +135,141 @@ conformal.pred.jack = function(x, y, x0, train.fun, predict.fun, alpha=0.1,
...
+                           nrow=n-1))
+    
+    return(one.lapply(1:m, function(l)  mad.train.fun(x[-i,,drop=F],r[,l])))
+     })
+    
+  }## mad.trained is a list  n x m
+  
+  
   # Compute leave-one-out residuals by brute-force
   else {
```
[the remark this answers is not in the document]
Comment (paolo-vergo) [reply conformalInference/R/jack.R:163]: Ok fixed

File: conformalInference/R/jack.R [hunk 5]
```diff
@@ -111,56 +135,141 @@ conformal.pred.jack = function(x, y, x0, train.fun, predict.fun, alpha=0.1,
...
+    
+    mad.trained<-one.lapply(1:n, function(i){ 
+    
+    r = abs(y[-i] - matrix(predict.fun(updated_models[[i]],x[-i,,drop=F]),
+                           nrow=n-1))
+    
+    return(one.lapply(1:m, function(l)  mad.train.fun(x[-i,,drop=F],r[,l])))
+     })
+    
+  }## mad.trained is a list  n x m
```
[the remark this answers is not in the document]
Comment (paolo-vergo) [reply conformalInference/R/jack.R]: Yeah I agree, it does not improve the performance. I will put it inside the loops

File: conformalInference/R/jack.R [hunk 6]
```diff
@@ -111,56 +135,141 @@ conformal.pred.jack = function(x, y, x0, train.fun, predict.fun, alpha=0.1,
...
+      mod_jj = train.fun(x[-jj,],y[-jj])
+      return(mod_jj)
+  })
+  
 
   # Compute leave-one-out residuals with special function
   if (!is.null(special.fun) &&
       (is.null(mad.train.fun) && is.null(mad.predict.fun))) {
     res = abs(special.fun(x,y,out))
   }
```
[the remark this answers is not in the document]
Comment (paolo-vergo) [reply conformalInference/R/jack.R]: I've considered this case!

**Comment by paolo-vergo** [comment]
@ryantibs I should have resolved all the conversations. 

As long as the list of people is concerned, they all contributed to the code (even if some were small contributions). Do you think, then, they should also  be listed as authors or rather as contributors ?
Let me know!

**Comment by ryantibs** [comment]
Thanks!

Re the authors in the description: whether author "aut" or contributor "ctb", I'll leave it completely up to you (I of course don't know the level of their contributions, since I only see your commits). You can read about the description of these two categories here: https://r-pkgs.org/description.html#description-authors

**Comment by paolo-vergo** [comment]
Ok! I'll read the description carefully and update it accordingly. 

Thanks!

**Comment by ryantibs** [comment]
Sorry for the delay. I just merged the changes now (after a typo fix and doc rebuild).

**Pull request merged**